test/randomized/test_transpiler_equivalence.py

```diff
@@ -0,0 +1,225 @@
+        self.qc.append(gate(*params), [qarg])
+
+    @rule(gate=st.sampled_from(twoQ_oneP_gates),
+          qargs=st.lists(qubits, max_size=2, min_size=2, unique=True),
+          param=st.floats(allow_nan=False, allow_infinity=False))
+    def add_2q1p_gate(self, gate, qargs, param):
+        """Append a random 2q gate with 1 random float parameter."""
+        self.qc.append(gate(param), qargs)
+
+    @rule(gate=st.sampled_from(twoQ_threeP_gates),
+          qargs=st.lists(qubits, max_size=2, min_size=2, unique=True),
+          params=st.lists(
+              st.floats(allow_nan=False, allow_infinity=False),
+              min_size=3, max_size=3))
+    def add_2q3p_gate(self, gate, qargs, params):
+        """Append a random 2q gate with 3 random float parameters."""
+        self.qc.append(gate(*params), qargs)
+
+    @rule(gate=st.sampled_from(oneQ_oneC_gates),
+          qarg=qubits,
+          carg=clbits)
+    def add_1q1c_gate(self, gate, qarg, carg):
+        """Append a random 1q, 1c gate."""
+        self.qc.append(gate(), [qarg], [carg])
+
+    @rule(gate=st.sampled_from(variadic_gates),
+          qargs=st.lists(qubits, min_size=1, unique=True))
+    def add_variQ_gate(self, gate, qargs):
+        """Append a gate with a variable number of qargs."""
+        self.qc.append(gate(len(qargs)), qargs)
+
+    @precondition(lambda self: len(self.qc.data) > 0)
+    @rule(carg=clbits,
+          data=st.data())
+    def add_c_if_last_gate(self, carg, data):
+        """Modify the last gate to be conditional on a classical register."""
+        creg = carg.register
+        val = data.draw(st.integers(min_value=0, max_value=2**len(creg)-1))
+
+        self.qc.data[-1][0].c_if(creg, val)
+
+    # Properties to check
+
+    @invariant()
+    def qasm(self):
+        """After each circuit operation, it should be possible to build QASM."""
+        self.qc.qasm()
+
+    @precondition(lambda self: any(isinstance(d[0], Measure) for d in self.qc.data))
+    @rule(
+        backend=st.one_of(
+            st.none(),
+            st.sampled_from(backends)),
+        opt_level=st.one_of(
+            st.none(),
+            st.integers(min_value=0, max_value=3)))
+    def equivalent_transpile(self, backend, opt_level):
+        """Simulate, transpile and simulate the present circuit. Verify that the
+        counts are not significantly different before and after transpilation.
+
+        """
+        shots = 4096
+
+        aer_qasm_simulator = Aer.get_backend('qasm_simulator')
+        aer_counts = execute(self.qc, backend=aer_qasm_simulator,
+                             shots=shots).result().get_counts()
+
+        xpiled_qc = transpile(self.qc, backend=backend, optimization_level=opt_level)
+        xpiled_aer_counts = execute(xpiled_qc, backend=aer_qasm_simulator,
+                                    shots=shots).result().get_counts()
+
+        assert _counts_equivalent(aer_counts, xpiled_aer_counts), \
```

Comment:
Since `assertDictAlmostEqual` is written to be a method on a `unittest.TestCase` object,
I can look into extracting out the comparison logic so we can re-use it here.

Reply:
Extracted message generation logic out of `assertDictAlmostEqual` in https://github.com/Qiskit/qiskit-terra/pull/2544/commits/bd4d91626ea7639c7bc69ea4607e814f98323225 . Still room to be cleaned up further though.